ng_slip: initial import

authmillenon opened the pull request:
Updated version of #1454. I removed the `board_uart0` support for now, to keep it simple.

~~Depends on #2575.~~ (merged)

~~Needs #2901 to build properly~~ (merged)

haukepetersen commented on sys/net/link_layer/ng_slip/ng_slip.c:
```diff
@@ -0,0 +1,293 @@
+                    break;
+
+                default:
+                    if (uart_write(uart, data[i]) < 0) {
```
this will not work (you will lose chars). Use `uart_write_blocking()` when you use the uart like this. Better (and I really would to this here): use the transmit interrupt and a transmit buffer!

```c
static ringbuffer_t *out_buf;

_slip_send(..) {
    ...
    default:
    ringbuffer_add_one(&out_buf, data[i]);
    uart_tx_start(uart);
    ...
}

// and
int _slip_rx_cb(void *arg) {
    if (_out_buf.avail > 0) {
        char c = (char)ringbuffer_get_one(&out_buf);
        uart_write(uart, c);
        return 1;
    }
    return 0;
}
```

authmillenon replied:
you mean tx_cb, right?

authmillenon replied:
The problem I have with this is, that we than need a second buffer of at least size `IPV6_MTU`.

haukepetersen replied:
yes

haukepetersen replied:
and true for the buffer size. But the blocking use of the UART interface is very slow. This basically means, that for sending one full IPv6 MTU, you controller basically blocks for about 10ms...

haukepetersen commented on sys/net/link_layer/ng_slip/ng_slip.c:
```diff
@@ -0,0 +1,293 @@
+static kernel_pid_t _slip_pid = KERNEL_PID_UNDEF;
+static uint32_t _bytes = 0;
+static uint8_t _esc = 0;
+static ringbuffer_t *_in_buf;
```
I don't like this. You should put all these parameters into a device descriptor and use that as a parameter with the functions below. Otherwise it is not possible to run more than one slip interface at the time.

haukepetersen commented:
You should spend the SLIP interface a device descriptor to make it possible to run more than one instance at a time! This could be interesting for setups, where one interface is connected to linux, and another one is connected to another RIOT based board - some of our boards have up to 8 UART interfaces :-)

authmillenon commented:
I was thinking about it, but than thought it might be an interesting challenge for future generations :D

haukepetersen commented:
hm, why not do it right in the first place? So others don't even get the idea to implement drivers with global variables (and the SLIP device would be in line with all new readio and device drivers...).

authmillenon commented:
You saw that I did it, right?

haukepetersen commented:
Ups, did not. But the new version does not work -> you will have data garbage as to shared uart buffers... I would really do it the same way we do any other device driver: declare the device descriptor in the header file and include the uart buffers with full size in this struct. Then it is up to the user of a slip device to allocate this device descriptor (not on the stack!) and just give the pointer to this memory to the init function.

authmillenon commented:
The uart buffers are not shared, they are uart local, but I can make the struct public. As for the full size data buffers: I chose very carefully not to include the size of the ringbuffers into this PR, simply because it mainly depends on the network layer what size is required.

authmillenon commented:
Done

authmillenon commented:
Sorry, somehow there was a foreign commit in my rebase

authmillenon commented:
Rebased to current master

authmillenon commented:
Rebased to current master

authmillenon commented:
Fixed a whitespace error

authmillenon commented:
Rebased to current master

haukepetersen commented on sys/include/net/ng_slip.h:
```diff
@@ -0,0 +1,88 @@
+    uint32_t in_bytes;      /**< the number of bytes received of a currently
+                             *   incoming packet */
+    uint16_t in_esc;        /**< receiver is in escape mode */
+    kernel_pid_t slip_pid;  /**< the PID of  preinitialized */
```
there slipped one space char in that is not supposed to be there...

haukepetersen commented on sys/include/net/ng_slip.h:
```diff
@@ -0,0 +1,88 @@
+ * @return  -ENOTSUP, if board does not implement @ref driver_periph_uart
+ * @return  -EOVERFLOW, if there are too many threads running already
+ */
+kernel_pid_t ng_slip_init(char priority, ng_slip_dev_t *dev, uint32_t baudrate);
```
you should add the uart as parameter here.

haukepetersen commented on sys/net/link_layer/ng_slip/ng_slip.c:
```diff
@@ -0,0 +1,285 @@
+kernel_pid_t ng_slip_init(char priority, ng_slip_dev_t *dev, uint32_t baudrate)
+{
+    (void)priority;
+    (void)uart;
```
uart not defined -> add as parameter to the init function.

haukepetersen commented on sys/net/link_layer/ng_slip/ng_slip.c:
```diff
@@ -0,0 +1,285 @@
+{
+    kernel_pid_t res;
+
+    if ((dev->uart < UART_0) || (dev->uart >= UART_NUMOF)) {
```
s/dev->uart/uart/

haukepetersen commented on sys/include/net/ng_slip.h:
```diff
@@ -0,0 +1,88 @@
+typedef struct {
+    uart_t uart;            /**< the UART interface */
+    ringbuffer_t *in_buf;   /**< input buffer */
+    ringbuffer_t *out_buf;  /**< output buffer */
```
also include the buffer memory into the file descriptor and initialize the ringbuffers inside the init() function.

haukepetersen commented on sys/net/link_layer/ng_slip/ng_slip.c:
```diff
@@ -0,0 +1,285 @@
+
+#define _SLIP_DEV(arg)    ((ng_slip_dev_t *)arg)
+
+static char _slip_stack[_SLIP_STACK_SIZE];
```
needs to be defined outside of this file and be a parameter to the init function -> otherwise you get in trouble once in initialize 2 slip devices...

haukepetersen commented:
With the code in the current state, it is still not possible to have more than 1 slip device... Needs some fixes

OlegHahm commented:
> You should spend the SLIP interface a device descriptor to make it possible to run more than one instance at a time! This could be interesting for setups, where one interface is connected to linux, and another one is connected to another RIOT based board - some of our boards have up to 8 UART interfaces :-)

Sorry, saw this only now, but actually I cannot think of any reasonable use case where I want to use more than one UART as a network device in the foreseeable future. Wouldn't it be wiser to spend more time with more pressing things inside the network stack and get slip support for **one** UART merged as soon as possible? (After all this PR is still the base for the default solution to turn any node into a border router, right?)

haukepetersen commented:
Disagree, what about testing?

haukepetersen commented:
and besides: if we decide to make it runnable only with one instance at a time, it should clearly reflect that in the code - and not something in between as it is now...

OlegHahm commented:
> Disagree, what about testing?

Testing what?

OlegHahm commented:
> and besides: if we decide to make it runnable only with one instance at a time, it should clearly reflect that in the code - and not something in between as it is now...

Yes, I guess now it's too late.

haukepetersen commented:
Testing the network stack? Slip is not only for communicating with Linux, but you can also connect simply two RIOT boards via UART. And the easiest test setup there is, is to take 2 jumper wires and connect 4 pins on a board, and voila, you have a full fledged network connection on one single board... For this reason alone I say it makes sense to have more than one possible slip device...

Further it is not much more work to enable multiple instances and it even forces some good structure to the code...

OlegHahm commented:
> Testing the network stack?

Don't we have nativenet exactly for this purpose? If I want to test the stack on a real device, I'm usually also interested in real transceivers. However, I agree that this discussion is kinda pointless now.

haukepetersen commented:
@authmillenon, should I just PR some changes against your branch?

authmillenon commented:
If you have some, yes please.

haukepetersen commented:
ok, will do

haukepetersen commented:
just did. Tell me what you think...

authmillenon commented:
Looks alright: merged it.

haukepetersen commented on tests/slip/Makefile:
```diff
@@ -0,0 +1,28 @@
+include ../Makefile.tests_common
+
+
+
```
you could remove two of the newlines here (sorry, that was probably me...).

haukepetersen commented:
nothing to add: ACK for the RIOT part, can't really say much about the Linux part...

authmillenon commented:
Rebased and squashed.

> nothing to add: ACK for the RIOT part, can't really say much about the Linux part...

Does that mean I need another ACK for that?

OlegHahm commented:
The linux part is imported from Contiki, right?

authmillenon commented:
Yes.

OlegHahm commented:
Then I don't think we need review. As long as it works it's good (at least for now).

kaspar030 commented:
+1

haukepetersen commented:
Agree, ACK then for everything.

gebart commented:
The Contiki tunslip6 program has some really ugly code, but it works, and I think it is better to import it as-is for now and improve it in follow-up PRs.

OlegHahm commented:
> The Contiki tunslip6 program has some really ugly code, but it works

Seems to be their motto. ;) *scnr*

OlegHahm commented:
However:
```
file has an unknown license header: 'dist/tools/tunslip/tapslip6.c'
file has an unknown license header: 'dist/tools/tunslip/tunslip.c'
file has an unknown license header: 'dist/tools/tunslip/tunslip6.c'
```

OlegHahm commented:
And the test application seems to fail for several boards (atmega2560, samr21, iot-lab_M3, and nucleo-f334).

authmillenon commented:
That's a bug in `auto_init_ng_netif`: It assumes that `ng_at86rfxx` is used as soon as `ng_netif` is build in. I guess this PR has to wait for #2901.

authmillenon commented:
> However:
> 
>     file has an unknown license header: 'dist/tools/tunslip/tapslip6.c'
>     file has an unknown license header: 'dist/tools/tunslip/tunslip.c'
>     file has an unknown license header: 'dist/tools/tunslip/tunslip6.c'

how do I teach the script the BSD 3-clause license?

OlegHahm commented:
Edit `dist/tools/licenses/patterns`.

authmillenon commented:
Done

authmillenon commented:
Rebased to #2901, added support for this module by it and adapted the test application for it.

authmillenon commented:
Somehow lost the changes to the test application. Anyhow: now they are in.

authmillenon commented:
Erm… does the ACK still uphold?

haukepetersen commented on tests/slip/main.c:
```diff
@@ -0,0 +1,76 @@
+    puts("SLIP test");
+
+    /* initialize and register pktdump */
+    dump.pid = ng_pktdump_init();
```
I think the syntactical correct call would be to `ng_pktdump_getpid()` here, but `init` works as well, up to you if you want to change this...

OlegHahm replied:
I think the `_getpid()` function is superfluous anyway.

haukepetersen replied:
I think its actually cleaner and more intuitive to have it, but I know you prefer to make the pid variable publicly accessible... :-)

authmillenon replied:
> I think the syntactical correct call would be to ng_pktdump_getpid() here, but init works as well, up to you if you want to change this...

To my defense: this function did not exist yet, when I wrote this line ;-P

haukepetersen commented:
ACK upholds, I leave it up to you if you address the comment above. Needs rebase and some squashing though...

authmillenon commented:
Squashed and rebased to master

authmillenon commented:
Depends on #3005 to be able to build on AVR.

authmillenon commented:
Rebased to current master and #3005.

authmillenon commented:
Another day another rebase :D

authmillenon commented:
Rebased to current master… can we FINALLY merge this (this still needs a re-ACK)

authmillenon commented:
Travis fails in unrelated applications. What's happening oO @phiros @OlegHahm @x3ro @gebart?

authmillenon commented:
Ah, a rebase error was the culprit. Let's see if it works now.

authmillenon commented:
This PR still fails due to the license checker… I added a pattern file which is to my knowledge correct, but it still fails… @LudwigOrtmann @OlegHahm @mehlis help!

authmillenon commented:
Okay… don't know exactly why, but it seems to be fixed... Now cppcheck complaining about the Contiki code… do I have to fix those?

gebart commented:
Ignore the cppcheck warnings/errors in tunslip6 (Contiki). We'll fix it in a follow up PR.

authmillenon commented:
Ok, then I'll merge

authmillenon commented:
\o/

OlegHahm commented on sys/include/net/ng_slip.h:
```diff
@@ -0,0 +1,93 @@
+typedef struct xbee_params {
+    uart_t uart;            /**< UART interfaced the device is connected to */
+    uint32_t baudrate;      /**< baudrate to use */
+} xbee_params_t;
```
Sorry to being **a bit** late: but this is copy pasta, right?

authmillenon replied:
Yes... Should I open a PR or did you already?

OlegHahm replied:
Nope, I didn't, so feel free.

OlegHahm commented:
Nope, I didn't, so feel free.
